### DISCUSSION TITLE
Object versioning support?

### DISCUSSION THREAD

**Discussion by cybercleric** [seaweedfs/seaweedfs#2627] [Ideas]
Hi, @chrislusf . I have looked at the code of this project and I want to participate in its improvement.
At the moment I don't see any functionality for versioning objects in the storage. Is such a solution implemented or planned in the future?
I can help with the implementation of such functionality.

[Example of a solution from Amazon](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html)

**Comment by chrislusf** [comment]
Thanks! Yes, versioning is not supported yet. Please have a design first.

**Reply by MercusChan** [reply]
This is amazing. 
As we known, *minio* supports object versioning as it claims.

What's the roadmap on support in *seaweedfs*?

**Comment by cybercleric** [comment]
Hi, Chris. This time I was debugging filler.sync. And I found a very bad bug. Synchronization of the modified file does not work correctly. This can be repeated:

Run two servers with volumes and fillers:
`server -dir=Server1alpha -master.port=11000 -filer -filer.port=11001 -volume.port=11002`
`server -dir=Server1sigma -master.port=11006 -filer -filer.port=11007 -volume.port=11008`

Run Active-Passive filler.sync:
`filer.sync -a localhost:11007 -b localhost:11001 -isActivePassive`

Upload file to 11007 port:
`curl -F file=@/Desktop/9.xml "http://localhost:11007/testFacebook/"`

If we request a file on two servers now, everything will be correct, even if we add data to the file and upload it again:
`curl "http://localhost:11007/testFacebook/9.xml"`
EQUALS
`curl "http://localhost:11001/testFacebook/9.xml"`

However, if we change the already existing data in the file (for example, we change the first line in the file, reducing its length), then this file on the second server will not be valid and will not be equivalent to the first file

<img width="819" alt="Снимок экрана 2022-02-07 в 14 21 11" src="https://user-images.githubusercontent.com/36068634/152779006-b3d2f983-a720-4b43-8b65-43ac9a14c262.png">

This problem occurs on **line 202 in the filer_sink.go** file. In particular, this is due to incorrect mapping of chunk names in the **DoMinusChunks function**. The names of deletedChunks do not match the chunks of existingEntry.Chunks, since the first chunks come from another server and have a different addressing (name) compared to the addressing on the server where the file is being overwritten.

Deleted chunks are not actually deleted on the server to which the file is replicated.

Now I do not know how to solve this problem, it looks quite complicated. However, the versioning mechanism that I'm going to implement should save us from this bug.

**Reply by chrislusf** [reply]
Thanks for reporting with these details and analyzing the problem!  I added a fix 9405eaefdbd8ee94dac9bd12e324e81afaa474f5

**Comment by Arshil-Akkala** [comment]
Hi, does seaweedfs support versioning yet?

**Reply by chrislusf** [reply]
Yes
